Add support for Schnorr crypto (BCH variant)

Fabcien opened the pull request:
Support Schnorr signature and verification defined in
https://github.com/bitcoincashorg/bitcoincash.org/blob/master/spec/2019-05-15-schnorr.md.

The implementation is based on https://github.com/Bitcoin-ABC/secp256k1
as well as the test vectors.

This is a first step toward adding support for Schnorr signature and verification to trezorlib.

onvej-sl commented:
Are you able to tell the difference between these signatures and Schnorr signatures defined in https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki, @Fabcien ?

Fabcien commented:
The goal is the same, but it differs in some choices being made:
 - The computation uses the 33 bytes compressed public key format (vs 32 bytes X only for BIP340)
 - The R.y is chosen as the one being a quadratic residue (vs the one being even for BIP340)

prusnak commented:
> The goal is the same, but it differs in some choices being made:

Would it be possible to rewrite your code in such a way it supports both BCH Schnorr and BIP-340 Schnorr? For example by adding `schnorr_type` enum with values `SCHNORR_BCH` and `SCHNORR_BIP340`

onvej-sl commented on crypto/schnorr.c:
```diff
@@ -0,0 +1,254 @@
+  bn_read_be(sign + 32, &s);
+
+  // Signature is invalid if s >= n or r >= p.
+  if (bn_is_zero(&r) || bn_is_zero(&s) || !bn_is_less(&r, &curve->prime) ||
```
Is (r != 0 and s != 0) actually needed? It's not written in the specification.

Fabcien replied:
Removed in bc6267c

onvej-sl commented on crypto/schnorr.c:
```diff
@@ -0,0 +1,254 @@
+
+int jacobi(const bignum256 *_n, const bignum256 *_k) {
+  int t = 0;
+  int k_is_one;
```
Please initialise _all_ local variables. It's our convention.

Fabcien replied:
Done in bc6267c

onvej-sl commented on crypto/schnorr.c:
```diff
@@ -0,0 +1,254 @@
+  bn_write_be(Rx, Rxbuf);
+
+  sha256_Init(&ctx);
+  sha256_Update(&ctx, Rxbuf, 32);
```
`sizeof(Rxbuf)` is better than 32

Fabcien replied:
Done in bc6267c

onvej-sl commented on crypto/schnorr.c:
```diff
@@ -0,0 +1,254 @@
+      t ^= (r == 3 || r == 5);
+    }
+
+    bn_cmov(&temp, 0, &n, &n);
+    bn_cmov(&n, 0, &k, &k);
+    bn_cmov(&k, 0, &temp, &temp);
```
Why don't you use `bn_copy` instead of `bn_cmov`?

Fabcien replied:
Done in bc6267c

onvej-sl commented on crypto/schnorr.c:
```diff
@@ -0,0 +1,254 @@
+
+  while (!bn_is_zero(&n)) {
+    while (bn_is_even(&n)) {
+      bn_rshift(&n);
```
You should comment a bit what is happening here. Something like:
jacobi(2 * n, k) = jacobi(n, k) if k == 1 (mod 8) or k == 7 (mod 8)
jacobi(2 * n, k) = -jacobi(n, k) if k == 3 (mod 8) or k == 5 (mod 8)

Fabcien replied:
Done in bc6267c

onvej-sl commented on crypto/schnorr.c:
```diff
@@ -0,0 +1,254 @@
+    bn_cmov(&n, 0, &k, &k);
+    bn_cmov(&k, 0, &temp, &temp);
+
+    t ^= ((n.val[0] & k.val[0] & 3) == 3);
```
```
jacobi(n, k) = jacobi(k, n) if k = n = 1 (mod 8)
jacobi(n, k) = -jacobi(k, n) if k = n = 3 (mod 8)
```

Fabcien replied:
Done in bc6267c

onvej-sl commented on crypto/schnorr.c:
```diff
@@ -0,0 +1,254 @@
+                          rfc6979_state *state) {
+  uint8_t hmac_data[SHA256_DIGEST_LENGTH + 16];
+
+  memcpy(hmac_data, hash, 32);
```
`SHA256_DIGEST_LENGTH` is better than 32.

Fabcien replied:
Done in bc6267c

onvej-sl commented on crypto/schnorr.c:
```diff
@@ -0,0 +1,254 @@
+  point_multiply(curve, &k, &curve->G, &R);
+
+  // If R.y is not a quadratic residue, negate the nonce
+  bn_normalize(&k);
```
I think `k` is already normalized. Maybe you wanted to reduce it?

Fabcien replied:
Neither is necessary, k is guaranteed to be < order. Removed in bc6267c

Fabcien commented:
@prusnak It can probably be done, but there are enough differences that it might be cleaner to just create new functions for BIP340. Code factorization can happen at this time to avoid duplication as needed.

prusnak commented:
> that it might be cleaner to just create new functions for BIP340

OK

andrewkozlik commented on crypto/schnorr.c:
```diff
@@ -0,0 +1,252 @@
+  // Compute R = sG - eP
+  bn_cnegate(1, &e, &curve->order);
+  bn_mod(&e, &curve->order);
+  point_multiply(curve, &s, &curve->G, &sG);
```
Use `scalar_multiply()`, it's specially written to be faster.
```suggestion
  scalar_multiply(curve, &s, &sG);
```

onvej-sl commented:
@Fabcien, please, leave conversation resolving to reviewers.

andrewkozlik commented on crypto/schnorr.c:
```diff
@@ -0,0 +1,252 @@
+  }
+
+  // Compute R = k * G
+  point_multiply(curve, &k, &curve->G, &R);
```
Use `scalar_multiply()`, see https://github.com/trezor/trezor-firmware/pull/1509/files#r589447096.

Fabcien replied:
Done in 1c7bd82

andrewkozlik commented on crypto/schnorr.c:
```diff
@@ -0,0 +1,252 @@
+  bn_copy(_n, &n);
+  bn_copy(_k, &k);
+
+  bn_mod(&n, &k);
```
Same comment as https://github.com/trezor/trezor-firmware/pull/1509/files#r588659565 applies to this `bn_mod` too, but in fact this `bn_mod` seems redundant, since `n < k` is guaranteed given that `_n` is always a point coordinate and `_k` is the order of the finite field. I'd just add `bn_is_less(_n, _k)` to the `assert` above.

onvej-sl replied:
> Same comment as https://github.com/trezor/trezor-firmware/pull/1509/files#r588659565 applies to this bn_mod too, but in fact this bn_mod seems redundant, since n < k is guaranteed given that _n is always a point coordinate and _k is the order of the finite field.

Yes, this is because the coordinates of a point that is an output of `scalar_multiply`, `point_multiply` or `point_add` are fully reduced.

> I'd just add bn_is_less(_n, _k) to the assert above.

I agree.

andrewkozlik replied:
Actually I take back my comment about adding the assert, because the algorithm works correctly even if the assertion doesn't hold, so it would be misleading to have it there. I just removed the `bn_mod()` in https://github.com/trezor/trezor-firmware/pull/1509/commits/2bf51f3a2c2b6e39e88f3face3e9a3be67a634ee.

onvej-sl commented on crypto/schnorr.c:
```diff
@@ -0,0 +1,252 @@
+
+  sha256_Init(&ctx);
+  sha256_Update(&ctx, Rxbuf, sizeof(Rxbuf));
+  sha256_Update(&ctx, pub_key, 33);
```
`sizeof(pub_key)` is better than `33`.

Fabcien replied:
I don't think this can work since `pub_key` is a parameter, this would be `sizeof(uint8_t *)`

onvej-sl replied:
Absolutely right.

onvej-sl commented on crypto/schnorr.c:
```diff
@@ -0,0 +1,252 @@
+  calc_e(curve, &r, pub_key, digest, &e);
+
+  // Compute R = sG - eP
+  bn_cnegate(1, &e, &curve->order);
```
> bn_subtract(&curve->order, &e, &e); seems more appropriate than bn_cnegate(), but I leave it to @onvej-sl to comment.

I totally agree.

onvej-sl replied:
> I am also wondering what to do if e comes out zero, in which case anyone can easily forge a signature. The likelihood of that happening by chance is infinitesimal, but it could be induced by a fault injection. I don't think we will be using Schnorr signatures in any situation where fault injections would be a concern, so I don't think we need to handle e==0, but I am asking for comments just in case anyone has a different opinion.

This is analogous to https://github.com/Fabcien/trezor-firmware/blob/add_schnorr_crypto/crypto/ecdsa.c#L1044_L1051. Is there a reason to handle it differently?

andrewkozlik replied:
We use ECDSA to verify signatures in the bootloader, where fault injections would be a concern, but I am not sure that we will be using Schnorr signatures for anything similar. On the other hand it wouldn't hurt to fail if `e==0`, so let's err on the side of caution and do so.

Fabcien replied:
I added the check in 1c7bd82

onvej-sl commented on crypto/schnorr.c:
```diff
@@ -0,0 +1,252 @@
+  bn_cnegate(is_non_quad_residue(&R.y, &curve->prime), &k, &curve->order);
+  bn_mod(&k, &curve->order);
+
+  bn_mod(&R.x, &curve->order);
```
You probably wanted `bn_mod(&R.x, &curve->prime)` but this is not necessary because the coordinates of the result of `scalar_multiply` (and `point_multiply`) are already fully reduced. (I agree it is badly documented).

Interesting is that no test failed but this is because `log2((prime - order) / prime)` is about `-128`, so it's very unlikely the coordinate was actually reduced.

Fabcien replied:
Removed in 1c7bd82

onvej-sl commented on crypto/schnorr.c:
```diff
@@ -0,0 +1,252 @@
+
+  // If R.y is not a quadratic residue, negate the nonce
+  bn_cnegate(is_non_quad_residue(&R.y, &curve->prime), &k, &curve->order);
+  bn_mod(&k, &curve->order);
```
This is true, since `k` goes to `bn_addmod` which cope with even unreduced numbers.

Fabcien replied:
Removed in 1c7bd82

onvej-sl commented on crypto/schnorr.c:
```diff
@@ -0,0 +1,252 @@
+  memzero(&private_key_scalar, sizeof(private_key_scalar));
+  bn_addmod(&s, &k, &curve->order);
+  memzero(&k, sizeof(k));
+  bn_fast_mod(&s, &curve->order);
```
Calling `bn_fast_most` is not necessary since an output of `bn_addmod` is partly reduced.

Fabcien replied:
Removed in 1c7bd82

onvej-sl commented on crypto/schnorr.c:
```diff
@@ -0,0 +1,252 @@
+    // jacobi(n, k) = -jacobi(k, n) if k = n = 3 (mod 4)
+    t ^= ((n.val[0] & k.val[0] & 3) == 3);
+
+    bn_mod(&n, &k);
```
I thought I commented it too, but I probably didn't...

The problem is that the function `bn_fast_mod(n, k)` you want to use works only for `2^256 - 2^224 <= k <= 2^256`. So we don't have a function for general reduction.

> so you might as well replace this line with bn_subtract(&n, &k, &n),

You can't, try `jacobi(7, 1)`.

I think the algorithm works correctly. My experimental results show that it needs about three times more iterations in the outer loop for 256-bit numbers. If the quadratic reciprocity is used only if `n < k` (so it either subtracts or uses the quadratic reciprocity) the algorithm needs only two times more iterations. That's not so bad.

andrewkozlik replied:
I never suggested using `bn_fast_mod()`. My point is that the `bn_mod()` here is highly misleading, because it will result in a different flow than one would expect and even when one realizes this, it's still far from obvious that the return value will be correct.

andrewkozlik replied:
Fixed in https://github.com/trezor/trezor-firmware/pull/1509/commits/2bf51f3a2c2b6e39e88f3face3e9a3be67a634ee. @onvej-sl please review and resolve.

onvej-sl replied:
> I never suggested using bn_fast_mod().

And I didn't mean it's you how want to use it. :)

onvej-sl replied:
Looks good!

andrewkozlik commented on crypto/schnorr.c:
```diff
@@ -0,0 +1,252 @@
+      t ^= (r == 3 || r == 5);
+    }
+
+    bn_copy(&n, &temp);
+    bn_copy(&k, &n);
+    bn_copy(&temp, &k);
```
Swapping pointers is more efficient than copying.

andrewkozlik replied:
Done in https://github.com/trezor/trezor-firmware/pull/1509/commits/2bf51f3a2c2b6e39e88f3face3e9a3be67a634ee.

Fabcien commented on crypto/schnorr.h:
```diff
@@ -0,0 +1,42 @@
+
+#include "ecdsa.h"
+
+/* A Schnorr signature is always 64 bits */
```
Good catch, done in 1c7bd82

Fabcien commented on crypto/schnorr.c:
```diff
@@ -0,0 +1,252 @@
+#include <stdio.h>
+#include <string.h>
+
+int jacobi(const bignum256 *_n, const bignum256 *_k) {
```
Done in 1c7bd82

Fabcien commented on crypto/schnorr.c:
```diff
@@ -0,0 +1,252 @@
+  if (generate_k_schnorr(curve, &k, &rng) != 0) {
+    memzero(&k, sizeof(k));
+    return 1;
+  }
```
I inlined `init_rfc6979_schnorr()` as suggested in 1c7bd82

Fabcien commented on crypto/schnorr.c:
```diff
@@ -0,0 +1,252 @@
+                 const uint8_t *msg, const uint32_t msg_len, uint8_t *sign) {
+  uint8_t digest[SHA256_DIGEST_LENGTH] = {0};
+
+  double_sha256(msg, msg_len, digest);
```
Removed in 1c7bd82

Fabcien commented on crypto/schnorr.c:
```diff
@@ -0,0 +1,252 @@
+  bn_copy(&e, &s);
+  bn_read_be(priv_key, &private_key_scalar);
+  bn_multiply(&private_key_scalar, &s, &curve->order);
+  memzero(&private_key_scalar, sizeof(private_key_scalar));
```
Nice ! Done in 1c7bd82

Fabcien commented:
Thanks for the great review! I hopefully addressed all your comments in 1c7bd82. This is a fairly large change, please let me know if you want me to squash the commits.

andrewkozlik reviewed (approved):
> Thanks for the great review! I hopefully addressed all your comments in 1c7bd82. This is a fairly large change, please let me know if you want me to squash the commits.

Good work @Fabcien. Thank you for your contribution. I already squashed the commits in https://github.com/trezor/trezor-firmware/tree/add_schnorr_crypto and we'll see if the CI tests pass: https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/pipelines/267555129. If they succeed, then I will force-push the squashed branch into this PR and we'll merge.

andrewkozlik commented:
CI tests successful. Ready to merge once @onvej-sl and @prusnak approve.

onvej-sl commented:
I approve (I'm not able to find the "Add review" button).

prusnak commented:
Thanks! Great work everyone!
